OPAL/si/src/main/scala/org/opalj/fpcf/scheduling/MaximumPhaseScheduling.scala

```diff
@@ -0,0 +1,235 @@
+                    } else {
+                        visited_batches = visited_batches :+ firstElement
+                        val keyList = aCyclicGraph.keys.toSet -- visited_batches
+                        if (keyList.nonEmpty) {
+                            aCyclicGraph = setLazyInAllBatchesInternal(aCyclicGraph, keyList.head)
+                        }
+                    }
+                } else {
+                    visited_batches = visited_batches :+ firstElement
+                    val keyList = aCyclicGraph.keys.toSet -- visited_batches
+                    if (keyList.nonEmpty) {
+                        setLazyInAllBatchesInternal(aCyclicGraph, keyList.head)
+                    }
+                }
+                aCyclicGraph
+            }
+            setLazyInAllBatchesInternal(aCyclicGraph, firstElement)
+        }
+
+        def topologicalSort(graph: Map[Int, List[Int]]): List[Int] = {
+            var sortedNodes: List[Int] = List.empty
+            var permanent: Set[Int] = Set.empty
+            var temporary: Set[Int] = Set.empty
+
+            val preparedGraph = graph.map { case (node, deps) =>
+                node -> deps.filter(_ != node)
+            }
+
+            def visit(node: Int): Unit = {
+                if (!permanent.contains(node)) {
+                    if (temporary.contains(node)) {
+                        throw new IllegalStateException("Graph contains a cycle")
+                    }
+                    temporary = temporary + node
+
+                    preparedGraph.get(node).head.foreach { otherNode => visit(otherNode) }
+
+                    permanent = permanent + node
+                    temporary = temporary - node
+
+                    sortedNodes = sortedNodes :+ node
+                }
+            }
+
+            for (node <- preparedGraph.keys) {
+                visit(node)
+            }
+            sortedNodes
+        }
+
+        computationSpecificationMap.foreach { csID =>
+            scheduleGraph += (csID._2 -> mapCSToNum(getAllCSFromPropertyBounds(csID._1.uses(ps))))
+        }
+
+        if (!scheduleLazyTransformerInAllBatches) {
+            scheduleGraph.foreach { node =>
+                if (computationSpecificationArray(node._1).computationType.equals(
+                        lazyType
+                    ) || computationSpecificationArray(node._1).computationType.equals(transformerType)
+                ) {
+                    scheduleGraph.foreach { subNode =>
+                        if (subNode._2.contains(node._1)) {
+                            scheduleGraph =
+                                scheduleGraph +
+                                    (node._1 -> (scheduleGraph.get(node._1).head ++ Set(subNode._1)))
+                        }
+                    }
+                }
+            }
+        }
+
+        var aCyclicGraph = sccs(scheduleGraph.size, edgeFunctionForSCCS)
```

Comment:
Is this "a cyclic graph"? If it is an "acyclic graph" instead, name it `acyclicGraph` to ensure understandability.